---

# Task 1: Familiarize Yourself with the Dataset

There are two main subtasks:  
1. Load the four data files that are part of this dataset.  
2. Explore each dataset and define the data types of each column. Describe the contents of each column, including the nature of the data values. Follow a similar approach as demonstrated in the lecture and tutorial. Discuss the shape of the data, value ranges, and possible categories. Convert any data into a more appropriate format if needed.  

Add as many code and markdown cells as necessary below. Typically, you should perform one type of investigation per code cell with output, and then comment on the results in a markdown cell below it.  

After completing the analysis, write a concise summary in the **Task 1: summary report** cell at the end. The final report should succinctly summarize the main characteristics of each data file, how they relate to each other, and any notable features.


In [12]:
# this is your first code cell, add more cells for code and mardown as needed
# imports
import pandas as pd
from IPython.display import display

In [ ]:
# TASK 1a: SESSIONS.CSV
df_sessions = pd.read_csv("data/sessions.csv")
display(df_sessions)
display (df_sessions.shape)
print(*df_sessions.columns, sep = ", ") # this prints the items in the list, separating them with a comma (https://www.geeksforgeeks.org/python/print-lists-in-python-4-different-ways/)

,session_id,viewer_id,streamer_id,stream_id,started_at,ended_at,duration_mins,chat_messages_sent,bits_cheered,followed_during,subscribed_during
0,SES000001,VWR00938,STR0055,SM00860,2024-01-18 19:16:00,2024-01-18 20:28:00,72,0,0,False,False
1,SES000002,VWR00804,STR0034,SM00542,2024-01-04 17:11:00,2024-01-04 19:41:00,150,11,0,False,False
2,SES000003,VWR00704,STR0008,SM00124,2024-02-15 15:11:00,2024-02-15 16:53:00,102,5,0,False,False
3,SES000004,VWR00926,STR0024,SM00373,2024-01-18 19:59:00,2024-01-18 20:56:00,57,3,9663,False,False
4,SES000005,VWR00883,STR0008,SM00120,2024-03-19 20:08:00,2024-03-19 20:22:00,14,7,0,False,False
...,...,...,...,...,...,...,...,...,...,...,...
22820,SES022821,VWR00856,STR0005,SM00083,2024-01-16 18:51:00,2024-01-16 22:17:00,206,31,0,False,False
22821,SES022822,VWR00985,STR0013,SM00200,2024-01-08 20:28:00,2024-01-08 20:52:00,24,1,0,False,False
22822,SES022823,VWR00645,STR0045,SM00711,2024-03-26 04:15:00,2024-03-26 04:58:00,43,1,0,False,False
22823,SES022824,VWR00291,STR0005,SM00078,2024-01-11 20:02:00,2024-01-11 21:01:00,59,2,0,False,False


(22825, 11)

session_id, viewer_id, streamer_id, stream_id, started_at, ended_at, duration_mins, chat_messages_sent, bits_cheered, followed_during, subscribed_during


<details>
<summary> Task 1a: click to expand </summary>

### Task 1a: Summarizing sessions.csv

Since every file is converted to a dataframe and is being treated like a table, I will often use the term table and file interchangably (within reason) in this assignment.

#### Overview
This table contains information about individual viewing sesssions.

#### Table shape
- **Rows**: There are 22825 rows (22824 entries, since one row is for the column headings.)
- **Columns**: There are 11 columns. They are described below.

##### Columns description
|Column Header | Decription | Type of data |
|--------------|------------|--------------|
| `session_id` | unique identifier for the inidividual session | Categorical (nominal) |
| `viewer_id` | unique identifer for the viewer (consistent across other files) | Cateogrical (nominal) |
| `streamer_id` | unique identifer for the viewer (consistent across other files) | Categorical (nominal) |
| `stream_id` | unique identifer for the stream session (consistent across other files) | Categorical (nominal) |
| `started_at` | Time and date the stream started. Format: YYYY-MM-DD HH:MM:SS | Numerical (interval) |
| `ended_at` | Time and date the stream ended. Format: YYYY-MM-DD HH:MM:SS | Numerical (interval) |
| `duration_mins` | The length of the stream in minutes. | Numerical (ratio) |
| `chat_messages_sent` | The number of messages set on the chat | Numerical (ratio) |
| `bits_cheered` | The number bit cheers sent (a twitch currency) | Numerical (ratio) |
| `followed_during` | Boolean representing if the viewer followed the streamer during the session | Categorical (nominal) |
| `subscribed_during` | Boolean representing if the viewer subscribed the streamer during the session | Categorical (nominal) |

</details>

In [ ]:
# TASK 1b: STREAMERS.CSV
df_streamers = pd.read_csv("data/streamers.csv")
display(df_streamers)
display (df_streamers.shape)
print(*df_streamers.columns, sep = ", ") # this prints the items in the list, separating them with a comma (https://www.geeksforgeeks.org/python/print-lists-in-python-4-different-ways/)

### Task 1 Summary Report

Write your report here, summarizing the work done in the cells above.

# Task 2: Data Cleaning, Diagnostics, and Corrections

Identify and resolve issues in the dataset. These can be of different kinds, such as:
- Look for duplicates or implausible values (e.g., users with negative followers, invalid age groups). Thoroughly document your exploration of the data. Decide whether these are outliers or the result of a systemic problem. If the latter, resolve the problem here; if they are outliers, we will address those later.  
- Define at least three heuristics (common-sense rules) to flag unusual entries (e.g., someone watching dozens of streams per day). Discuss whether what you found are errors or artifacts.  

I will repeat the instructions here, but this is expected for each task: each code cell should focus on a specific analysis. You should run the code, and in the markdown below each code cell and its result, provide a detailed interpretation of what you have uncovered in that step. After completing the analyses, summarize the key findings in the indicated summary report cell below.


In [2]:
# start code for Task 2 here

### Task 2 summary report

# Task 3: Descriptive Statistics and Visualizations

- Create summary statistics and visualizations to understand user characteristics  
- Create summary statistics for streams data overall, and summaries by age group, account tire, or peak viewers  
- Visualize the distribution of streams durations, popular categories,   
- Show category-specific information and compare streamers based on categories  

In [3]:
#start code for Task 3 here

### Task 3 Summary report
summarize your findings here. 

# Task 4: Dealing with Outliers and Missing Values

- Are there any missing values or outliers in the original data? Perform the analysis and report your findings.  
- Are there any outliers or missing values in the derived data? Perform the analysis and report your findings.  
- If there are no outliers or missing values, introduce 3–5% of missing values in a selected column and 1–2% of outliers in another column.  
- Experiment with two strategies for imputing missing values and two strategies for handling outliers. Show how each strategy affects the characteristics of the data.


In [4]:
#start code for Task 4 here

### Task 4 Summary report

# Task 5: Correlation and Behavioral Patterns

- Compute pairwise correlations between user features and the number of check-ins  
- Visualize the results using a correlation matrix (use a coolwarm color scheme and proper scaling)  
- Try Spearman rank correlation and explain when it is more appropriate than Pearson

In [5]:
# tast 5 code starts here

### Task 5 Summary report

# Task 6: Compare Behavior Across User Segments

- Segment users into 2–3 groups (e.g., by age bracket, account tire)  
- Compare their viewing behavior over time using boxplots or dot plots  
- Identify any outliers in group behavior and investigate them

In [6]:
# task 6 code start here

### Task 6 Summary report

# Task 7: Your Own EDA Questions

Formulate and explore two hypotheses of your own:  
- Propose questions based on curiosity (e.g., “Does a user's variety of categories influence morning vs. evening or session duration?”)  
- Use summary statistics and visualizations to answer each question  
- Briefly discuss what additional data would help you answer the questions more effectively


In [7]:
# task 7 code start here

### Task 7 Summary report

# Task 8: EDA Presentation (Insight Report)

Summarize your findings in a polished format:  
- Create 3–4 slides or a single-page report highlighting your most important insights  
- Include at least 2 charts and interpret them using clear, plain language  
- Apply Tufte’s principles: minimize chartjunk, maximize data-to-ink ratio, and ensure clear labeling  